Code/src/main/java/nl/utwente/viskell/ui/components/Block.java

```diff
@@ -160,37 +161,47 @@ public void handleConnectionChanges() {
             }
         }
 
+        // Now the change is not propagated any further start recomputation.
         if (!propagatedDown) {
-            // Now the change is not propagated any further, start type checking the dirty expressions.
-            try {
-                // Analyze the entire tree.
-                this.getExpr().findType();
-                getPane().setErrorOccurred(false);
-                // TODO: This will set the errorOccurred for the entire
-                // program, not just the invalidated tree. This means that
-                // when having multiple small program trees, errors get
-                // reset to quickly.
-
-                // No type mismatches.
-            } catch (HaskellTypeError e) {
-                // A Type mismatch occurred.
-                int index = -1;
-                // Determine the input index of the Type error.
-                Expression errorExpr = e.getExpression();
-                while (errorExpr instanceof Apply) {
-                    errorExpr = ((Apply) errorExpr).getChildren().get(0);
-                    index++;
-                }
-                // Get the Block in which the type error occurred and
-                // set the error state for the mismatched input to true.
-                getPane().getExprToFunction(errorExpr).getInput(index).setErrorState(true);
-                // Indicate that an error occurred in the latest analyze attempt.
-                getPane().setErrorOccurred(true);
-            }
+            // Needs to be delayed, because recomputation clears exprIsDirty also used to avoid infinite recursion.
+            Platform.runLater(this::recomputeExpression);
```

Comment:
All connected blocks get marked dirty immediately in handleConnectionChanges, and as long as runLater is the same gui thread it will be fine.
This was the least ugly solution i could find...